akka-projection-testkit/src/main/scala/akka/projection/testkit/scaladsl/TestProjection.scala

```diff
@@ -0,0 +1,114 @@
+      sourceProvider: SourceProvider[Offset, EventEnvelope[Event]],
+      handler: Handler[EventEnvelope[Event]]): Projection[EventEnvelope[Event]] =
+    new TestProjection(projectionId, sourceProvider, handler)(system)
+}
+
+class TestProjection[Event](
+    val projectionId: ProjectionId,
+    sourceProvider: SourceProvider[Offset, EventEnvelope[Event]],
+    handler: Handler[EventEnvelope[Event]])(implicit val system: ActorSystem[_])
+    extends Projection[EventEnvelope[Event]]
+    with SettingsImpl[TestProjection[Event]] {
+
+  override val statusObserver: StatusObserver[EventEnvelope[Event]] = NoopStatusObserver
+
+  override def withStatusObserver(observer: StatusObserver[EventEnvelope[Event]]): TestProjection[Event] =
+    this // no need for StatusObserver in tests
+
+  final override def withRestartBackoffSettings(restartBackoff: RestartBackoffSettings): TestProjection[Event] = this
+  override def withSaveOffset(afterEnvelopes: Int, afterDuration: FiniteDuration): TestProjection[Event] = this
+  override def withGroup(groupAfterEnvelopes: Int, groupAfterDuration: FiniteDuration): TestProjection[Event] = this
+
+  private[akka] def actorHandlerInit[T]: Option[ActorHandlerInit[T]] = None
+
+  override def run()(implicit system: ActorSystem[_]): RunningProjection =
+    new InternalProjectionState(sourceProvider, handler).newRunningInstance()
+
+  private[projection] def mappedSource()(implicit system: ActorSystem[_]): Source[Done, Future[Done]] =
+    new InternalProjectionState(sourceProvider, handler).mappedSource()
+
+  /*
+   * INTERNAL API
+   * This internal class will hold the KillSwitch that is needed
+   * when building the mappedSource and when running the projection (to stop)
+   */
+  private class InternalProjectionState(
+      sourceProvider: SourceProvider[Offset, EventEnvelope[Event]],
+      handler: Handler[EventEnvelope[Event]])(implicit val system: ActorSystem[_]) {
+
+    private val killSwitch = KillSwitches.shared(projectionId.id)
+
+    def mappedSource(): Source[Done, Future[Done]] =
+      Source
+        .futureSource(sourceProvider
+          .source(null))
+        .via(killSwitch.flow)
+        .mapAsync(1)(i => handler.process(i))
+        .mapMaterializedValue(_ => Future.successful(Done))
+
+    def newRunningInstance(): RunningProjection =
+      new TestRunningProjection(mappedSource(), killSwitch)
+  }
+
+  private class TestRunningProjection(val source: Source[Done, _], killSwitch: SharedKillSwitch)
+      extends RunningProjection {
+
+    private val futureDone = source.run()
+
+    override def stop(): Future[Done] = {
+      killSwitch.shutdown()
+      futureDone
+    }
+  }
+}
+
+// FIXME: this should be replaced as part of #198
+object TestSourceProvider {
+  def apply[Event](sourceEvents: List[(Long, Event)]): SourceProvider[Offset, EventEnvelope[Event]] = {
+    new TestSourceProvider[Event](sourceEvents)
+  }
+}
+
+class TestSourceProvider[Event](sourceEvents: List[(Long, Event)])
```

Comment:
I think this would be better as `sourceEvents: immutable.Seq[EventEnvelope]` so that the user can define all things for the EventEnvelope.

Name should be EventSourcedProvider, corresponding to EventSourcedProvider.

Reply:
> Name should be EventSourcedProvider, corresponding to EventSourcedProvider.

I'm not sure I understand.  Are you suggesting `TestSourceProvider` be renamed to `EventSourcedProvider`?  Since the user now defines the envelope it may not actually use the akka event sourced envelope.